File: core/src/main/java/kandroid/observer/kcsapi/api_start2.java
package kandroid.observer.kcsapi;

import com.google.gson.Gson;

import java.util.List;

import kandroid.data.KCDatabase;
import kandroid.data.Start2Data;
import kandroid.observer.ApiBase;
import kandroid.observer.RawData;

public class api_start2 extends ApiBase {

    public static final String API_NAME = "api_start2";

    @Override
    protected void onDataReceived(RawData rawData) {
        ApiStart2 apiStart2 = new Gson().fromJson(rawData.toString(), ApiStart2.class);
        Start2Data start2Data = KCDatabase.INSTANCE.getMaster();

        // api_mst_ship
        for (ApiStart2.ApiData.ApiMstShip apiMstShip : apiStart2.api_data.api_mst_ship) {
            int id = apiMstShip.api_id;
            Start2Data.MasterShipData ship = start2Data.masterShipData.get(id);
            if (ship == null) {
                ship = new Start2Data.MasterShipData();
                ship.setData(apiMstShip);
                start2Data.masterShipData.put(ship);
            } else {
                ship.setData(apiMstShip);
            }
        }

        // api_mst_slotitem_equiptype
        for (ApiStart2.ApiData.ApiMstSlotitemEquiptype apiMstSlotitemEquiptype : apiStart2.api_data.api_mst_slotitem_equiptype) {
            int id = apiMstSlotitemEquiptype.api_id;
            Start2Data.EquipmentType type = start2Data.equipmentType.get(id);
            if (type == null) {
                type = new Start2Data.EquipmentType();
                type.setData(apiMstSlotitemEquiptype);
                start2Data.equipmentType.put(type);
            } else {
                type.setData(apiMstSlotitemEquiptype);
            }
        }

        // api_mst_stype 特別置換処理 mdzz
        apiStart2.api_data.api_mst_stype.get(7).api_name = "巡洋戦艦";
        for (ApiStart2.ApiData.ApiMstStype apiMstStype : apiStart2.api_data.api_mst_stype) {
            int id = apiMstStype.api_id;
            Start2Data.ShipType type = start2Data.shipType.get(id);
            if (type == null) {
                type = new Start2Data.ShipType();
                type.setData(apiMstStype);
                start2Data.shipType.put(type);
            } else {
                type.setData(apiMstStype);
            }
        }

        // api_mst_slotitem
        for (ApiStart2.ApiData.ApiMstSlotitem apiMstSlotitem : apiStart2.api_data.api_mst_slotitem) {
            int id = apiMstSlotitem.api_id;
            Start2Data.MasterEquipmentData equipment = start2Data.masterEquipmentData.get(id);
            if (equipment == null) {
                equipment = new Start2Data.MasterEquipmentData();
                equipment.setData(apiMstSlotitem);
                start2Data.masterEquipmentData.put(equipment);
            } else {
                equipment.setData(apiMstSlotitem);
            }
        }

        // api_mst_useitem
        for (ApiStart2.ApiData.ApiMstUseitem apiMstUseitem : apiStart2.api_data.api_mst_useitem) {
            int id = apiMstUseitem.api_id;
            Start2Data.MasterUseItemData useItem = start2Data.masterUseItemData.get(id);
            if (useItem == null) {
                useItem = new Start2Data.MasterUseItemData();
                useItem.setData(apiMstUseitem);
                start2Data.masterUseItemData.put(useItem);
            } else {
                useItem.setData(apiMstUseitem);
            }
        }

        // api_mst_mapinfo
        for (ApiStart2.ApiData.ApiMstMapinfo apiMstMapinfo : apiStart2.api_data.api_mst_mapinfo) {
            int id = apiMstMapinfo.api_id;
            Start2Data.MasterMapInfoData mapInfo = start2Data.masterMapInfoData.get(id);
            if (mapInfo == null) {
                mapInfo = new Start2Data.MasterMapInfoData();
                mapInfo.setData(apiMstMapinfo);
                start2Data.masterMapInfoData.put(mapInfo);
            } else {
                mapInfo.setData(apiMstMapinfo);
            }
        }

        // api_mst_mission
        for (ApiStart2.ApiData.ApiMstMission apiMstMission : apiStart2.api_data.api_mst_mission) {
            int id = apiMstMission.api_id;
            Start2Data.MasterMissionData mission = start2Data.masterMissionData.get(id);
            if (mission == null) {
                mission = new Start2Data.MasterMissionData();
                mission.setData(apiMstMission);
                start2Data.masterMissionData.put(mission);
            } else {
                mission.setData(apiMstMission);
            }
        }

        // TODO api_mst_shipupgrade
//        for (api_start2.ApiStart2.ApiData.ApiMstShipupgrade apiMstShipupgrade : apiStart2.api_data.api_mst_shipupgrade) {
//            int idbefore = apiMstShipupgrade.api_current_ship_id;
//            int idafter = apiMstShipupgrade.api_id;
//            MasterShipData shipbefore = masterShipData.getApi(idbefore);
//            MasterShipData shipafter = masterShipData.getApi(idafter);
//            int level = apiMstShipupgrade.api_upgrade_level;
//        }

        KCDatabase.INSTANCE.saveMaster();
    }

    @Override
    public String getApiName() {
        return API_NAME;
    }

    public static class ApiStart2 {

        public ApiData api_data;
        public int api_result;
        public String api_result_msg;

        public static class ApiData {
            public ApiMstConst api_mst_const;
            public ApiMstItemShop api_mst_item_shop;
            public List<ApiMstBgm> api_mst_bgm;
            public List<Integer> api_mst_equip_exslot;
            public List<ApiMstFurniture> api_mst_furniture;
            public List<ApiMstFurnituregraph> api_mst_furnituregraph;
            public List<ApiMstMaparea> api_mst_maparea;
            public List<ApiMstMapbgm> api_mst_mapbgm;
            public List<ApiMstMapcell> api_mst_mapcell;
            public List<ApiMstMapinfo> api_mst_mapinfo;
            public List<ApiMstMission> api_mst_mission;
            public List<ApiMstPayitem> api_mst_payitem;
            public List<ApiMstShip> api_mst_ship;
            public List<ApiMstShipgraph> api_mst_shipgraph;
            public List<ApiMstShipupgrade> api_mst_shipupgrade;
            public List<ApiMstSlotitem> api_mst_slotitem;
            public List<ApiMstSlotitemEquiptype> api_mst_slotitem_equiptype;
            public List<ApiMstStype> api_mst_stype;
            public List<ApiMstUseitem> api_mst_useitem;

            public static class ApiMstConst {
                public ApiBokoMaxShips api_boko_max_ships;
                public ApiDpflagQuest api_dpflag_quest;
                public ApiParallelQuestMax api_parallel_quest_max;

                public static class ApiBokoMaxShips {
                    public int api_int_value;
                    public String api_string_value;
                }

                public static class ApiDpflagQuest {
                    public int api_int_value;
                    public String api_string_value;
                }

                public static class ApiParallelQuestMax {
                    public int api_int_value;
                    public String api_string_value;
                }
            }

            public static class ApiMstItemShop {
                public List<Integer> api_cabinet_1;
                public List<Integer> api_cabinet_2;
            }

            public static class ApiMstBgm {
                public int api_id;
                public String api_name;
            }

            public static class ApiMstFurniture {
                public String api_description;
                public int api_id;
                public int api_no;
                public int api_price;
                public int api_rarity;
                public int api_saleflg;
                public int api_season;
                public String api_title;
                public int api_type;
            }

            public static class ApiMstFurnituregraph {
                public String api_filename;
                public int api_id;
                public int api_no;
                public int api_type;
                public String api_version;
            }

            public static class ApiMstMaparea {
                public int api_id;
                public String api_name;
                public int api_type;
            }

            public static class ApiMstMapbgm {
                public int api_id;
                public int api_maparea_id;
                public int api_no;
                public List<Integer> api_boss_bgm;
                public List<Integer> api_map_bgm;
            }

            public static class ApiMstMapcell {
                public int api_color_no;
                public int api_id;
                public int api_map_no;
                public int api_maparea_id;
                public int api_mapinfo_no;
                public int api_no;
            }

            public static class ApiMstMapinfo extends Start2Data.MasterMapInfoData.ApiMstMapinfo {
            }

            public static class ApiMstMission extends Start2Data.MasterMissionData.ApiMstMission {
            }

            public static class ApiMstPayitem {
                public String api_description;
                public int api_id;
                public String api_name;
                public int api_price;
                public int api_type;
                public List<Integer> api_item;
            }

            public static class ApiMstShip extends Start2Data.MasterShipData.ApiMstShip {
            }

            public static class ApiMstShipgraph {
                public String api_filename;
                public int api_id;
                public int api_sortno;
                public List<Integer> api_battle_d;
                public List<Integer> api_battle_n;
                public List<Integer> api_boko_d;
                public List<Integer> api_boko_n;
                public List<Integer> api_ensyue_n;
                public List<Integer> api_ensyuf_d;
                public List<Integer> api_ensyuf_n;
                public List<Integer> api_kaisyu_d;
                public List<Integer> api_kaisyu_n;
                public List<Integer> api_kaizo_d;
                public List<Integer> api_kaizo_n;
                public List<Integer> api_map_d;
                public List<Integer> api_map_n;
                public List<String> api_version;
                public List<Integer> api_weda;
                public List<Integer> api_wedb;
            }

            public static class ApiMstShipupgrade {
                public int api_catapult_count;
                public int api_current_ship_id;
                public int api_drawing_count;
                public int api_id;
                public int api_original_ship_id;
                public int api_sortno;
                public int api_upgrade_level;
                public int api_upgrade_type;
            }

            public static class ApiMstSlotitem extends Start2Data.MasterEquipmentData.ApiMstSlotitem {
            }

            public static class ApiMstSlotitemEquiptype extends Start2Data.EquipmentType.ApiMstSlotitemEquiptype {
            }

            public static class ApiMstStype extends Start2Data.ShipType.ApiMstStype {
            }

            public static class ApiMstUseitem extends Start2Data.MasterUseItemData.ApiMstUseitem {
            }
        }
    }
}

File: core/src/main/java/kandroid/observer/kcsapi/api_start2.kt
package kandroid.observer.kcsapi

import kandroid.config.Config
import kandroid.config.FILE_START2
import kandroid.data.*
import kandroid.observer.ApiBase
import kandroid.observer.RawData
import kandroid.utils.collection.SparseIntArray
import kandroid.utils.json.*
import org.apache.commons.io.FileUtils
import java.nio.charset.Charset

object api_start2 : ApiBase() {
    override val name: String get() = "api_start2"

    override fun onDataReceived(rawData: RawData) {
        // 保存Start2以便下次载入
        if (rawData.isFromKcServer) {
            val file = Config.getSaveUserDataFile(FILE_START2)
            FileUtils.writeStringToFile(file, rawData.responseString, Charset.defaultCharset())
        }

        val data = rawData["api_data"].obj ?: return

        val master = KCDatabase.Master
        // 特別置換処理
        data["api_mst_stype"][7]["api_name"] = "巡洋戦艦"

        // api_mst_ship
        val api_mst_ship = data["api_mst_ship"].array
        if (api_mst_ship != null) {
            for (elem in api_mst_ship) {
                val id = elem["api_id"].int()
                var ship = master.ship[id]
                if (ship == null) {
                    ship = MasterShipData()
                    ship.loadFromResponse(name, elem)
                    master.ship.put(ship)
                } else {
                    ship.loadFromResponse(name, elem)
                }
            }
        }

        // 改装関連のデータ設定
        for (ship in master.ship) {
            val remodelID = ship.remodelAfterShipId
            if (remodelID != 0) {
                master.ship[remodelID]?.remodelBeforeShipId = ship.shipId
            }
        }

        // TODO api_mst_shipgraph

        // api_mst_slotitem_equiptype
        val api_mst_slotitem_equiptype = data["api_mst_slotitem_equiptype"].array
        if (api_mst_slotitem_equiptype != null) {
            for (elem in api_mst_slotitem_equiptype) {
                val id = elem["api_id"].int()
                var equipmentType = master.equipmentType[id]
                if (equipmentType == null) {
                    equipmentType = EquipmentType()
                    equipmentType.loadFromResponse(name, elem)
                    master.equipmentType.put(equipmentType)
                } else {
                    equipmentType.loadFromResponse(name, elem)
                }
            }
        }

        // api_mst_stype
        val api_mst_stype = data["api_mst_stype"].array
        if (api_mst_stype != null) {
            for (elem in api_mst_stype) {
                val id = elem["api_id"].int()
                var shipType = master.shipType[id]
                if (shipType == null) {
                    shipType = ShipType()
                    shipType.loadFromResponse(name, elem)
                    master.shipType.put(shipType)
                } else {
                    shipType.loadFromResponse(name, elem)
                }
            }
        }

        // api_mst_slotitem
        val api_mst_slotitem = data["api_mst_slotitem"].array
        if (api_mst_slotitem != null) {
            for (elem in api_mst_slotitem) {
                val id = elem["api_id"].int()
                var equipment = master.equipment[id]
                if (equipment == null) {
                    equipment = MasterEquipmentData()
                    equipment.loadFromResponse(name, elem)
                    master.equipment.put(equipment)
                } else {
                    equipment.loadFromResponse(name, elem)
                }
            }
        }

        // api_mst_useitem
        val api_mst_useitem = data["api_mst_useitem"].array
        if (api_mst_useitem != null) {
            for (elem in api_mst_useitem) {
                val id = elem["api_id"].int()
                var useItem = master.useItem[id]
                if (useItem == null) {
                    useItem = MasterUseItemData()
                    useItem.loadFromResponse(name, elem)
                    master.useItem.put(useItem)
                } else {
                    useItem.loadFromResponse(name, elem)
                }
            }
        }

        // api_mst_maparea
        val api_mst_maparea = data["api_mst_maparea"].array
        if (api_mst_maparea != null) {
            for (elem in api_mst_maparea) {
                val id = elem["api_id"].int()
                var mapArea = master.mapArea[id]
                if (mapArea == null) {
                    mapArea = MasterMapAreaData()
                    mapArea.loadFromResponse(name, elem)
                    master.mapArea.put(mapArea)
                } else {
                    mapArea.loadFromResponse(name, elem)
                }
            }
        }

        // api_mst_mapinfo
        val api_mst_mapinfo = data["api_mst_mapinfo"].array
        if (api_mst_mapinfo != null) {
            for (elem in api_mst_mapinfo) {
                val id = elem["api_id"].int()
                var mapInfo = master.mapInfo[id]
                if (mapInfo == null) {
                    mapInfo = MasterMapInfoData()
                    mapInfo.loadFromResponse(name, elem)
                    master.mapInfo.put(mapInfo)
                } else {
                    mapInfo.loadFromResponse(name, elem)
                }
            }
        }

        // api_mst_mission
        val api_mst_mission = data["api_mst_mission"].array
        if (api_mst_mission != null) {
            for (elem in api_mst_mission) {
                val id = elem["api_id"].int()
                var mapInfo = master.mission[id]
                if (mapInfo == null) {
                    mapInfo = MasterMissionData()
                    mapInfo.loadFromResponse(name, elem)
                    master.mission.put(mapInfo)
                } else {
                    mapInfo.loadFromResponse(name, elem)
                }
            }
        }

        // api_mst_shipupgrade
        val api_mst_shipupgrade = data["api_mst_shipupgrade"].array
        if (api_mst_shipupgrade != null) {
            val upgradeTimes = SparseIntArray(70) // 记录本船id和第几次改造
            for (elem in api_mst_shipupgrade) {
                val idAfter = elem["api_id"].int() // 本船的id
                val idBefore = elem["api_current_ship_id"].int() // 本船改造前的id
                val level = elem["api_upgrade_level"].int() // 这是本船第几次改造
                val shipAfter = master.ship[idAfter]

                // 例：太太改二咸(api_id=461)可以由太太改(id=288)和太太改二甲(id=466)改造而来
                val valueIfKeyNotFound = -1
                val l = upgradeTimes[idAfter, valueIfKeyNotFound]
                if (l != valueIfKeyNotFound) { // 即upgradeLevels.containsKey(idAfter)
                    if (level < l) { //
                        shipAfter?.remodelBeforeShipId = idBefore
                        upgradeTimes.put(idAfter, level)
                    }
                } else { //
                    shipAfter?.remodelBeforeShipId = idBefore
                    upgradeTimes.put(idAfter, level)
                }

                val shipBefore = master.ship[idBefore]
                if (shipBefore != null) {
                    shipBefore.needBlueprint = elem["api_drawing_count"].int()
                    shipBefore.needCatapult = elem["api_catapult_count"].int()
                }
            }
        }
    }
}